File: server/matrix_operations.py
import numpy as np

def add(a: np.ndarray, b: np.ndarray) -> np.ndarray:
    return np.add(a, b)

def subtract(a: np.ndarray, b: np.ndarray) -> np.ndarray:
    return np.subtract(a, b)

def transpose(a: np.ndarray) -> np.ndarray:
    return np.transpose(a)

def dot(a: np.ndarray, b: np.ndarray) -> np.ndarray:
    return np.dot(a, b)

def cross(a: np.ndarray, b: np.ndarray) -> np.ndarray:
    return np.cross(a, b)

def inverse(a: np.ndarray) -> np.ndarray:
    return np.linalg.inv(a)

def det(a: np.ndarray) -> np.ndarray:
    return np.array([[np.linalg.det(a)]])

File: src/components/utils/serverless/MatrixOperations.js
import * as math from 'mathjs';
import { Matrix } from '../Matrix.js';

export class MatrixOperations
{
  static add(a, b)
  {
    const data = math.add(
      math.reshape(a.values, [a.rows, a.columns]),
      math.reshape(b.values, [b.rows, b.columns])
    );

    return new Matrix(
      data.flat(), 
      ...math.size(data)
    );
  }

  static subtract(a, b)
  {
    const data = math.subtract(
      math.reshape(a.values, [a.rows, a.columns]),
      math.reshape(b.values, [b.rows, b.columns])
    );

    return new Matrix(
      data.flat(), 
      ...math.size(data)
    );
  }

  static transpose(m)
  {
    const data = math.transpose(
      math.reshape(m.values, [m.rows, m.columns])
    );


    return new Matrix(
      data.flat(), 
      ...math.size(data)
    );
  }

  static dot(a, b)
  {
    const data = math.multiply(
      math.reshape(a.values, [a.rows, a.columns]),
      math.reshape(b.values, [b.rows, b.columns])
    );

    return new Matrix(
      typeof data === 'number' ? [data] : data.flat(), 
      ...math.size(data)
    );
  }

  static cross(a, b)
  {
    const data = math.cross(
      math.reshape(a.values, [a.rows, a.columns]),
      math.reshape(b.values, [b.rows, b.columns])
    );

    return new Matrix(
      data.flat(), 
      ...math.size(data)
    );
  }

  static inverse(m) 
  {
    const data = math.inv(
      math.reshape(m.values, [m.rows, m.columns])
    );

    return new Matrix(
      data.flat(), 
      ...math.size(data)
    );
  }

  static det(m) 
  {
    const data = math.det(
      math.reshape(m.values, [m.rows, m.columns])
    );

    return new Matrix(
      [data], 
      1, 1
    );
  }
}